File: C++/12.享元模式/FlyweightPatternDemo.cpp
/**
 * @file FlyweightPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-12
 *
 * @copyright Copyright (c) 2019
 *
 */

#include <stdlib.h>

#include "ShapeFactory.h"
#define random(x) rand() % (x)

static std::string colors[] = {"Red", "Green", "Blue", "White", "Black"};
static std::string& GetRandomColor() { return colors[(int)(random(sizeof(colors) / sizeof(colors[0])))]; }
static int GetRandomX() { return (int)random(100); }
static int GetRandomY() { return (int)random(100); }

int main(int argc, char const* argv[]) {
    for (int i = 0; i < 20; ++i) {
        std::shared_ptr<Circle> circle = ShapeFactory::GetCircle(GetRandomColor());
        circle->SetX(GetRandomX());
        circle->SetY(GetRandomY());
        circle->SetRadius(100);
        circle->Draw();
    }
    return 0;
}

File: Objective-C/12.FlyweightPattern/FlyweightPatternDemo.m
//
//  FlyweightPatternDemo.m
//  12.FlyweightPattern
//
//  Created by Layne on 2020/10/18.
//

#import "Circle.h"
#import "ShapeFactory.h"

static NSArray* colors[3] = {@"one", @"two", @"three"};

static NSString* getRandomColor() { return colors[random() % (sizeof(colors) / sizeof(NSArray*))]; }
static int getRandomX() { return (random() % 100); }
static int getRandomY() { return (random() % 100); }

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        NSLog(@"Hello World!.");
        for (int i = 0; i < 20; ++i) {
            Circle* circle = [ShapeFactory getCircle:getRandomColor()];
            [circle setX:getRandomX()];
            [circle setY:getRandomY()];
            [circle setRadius:100];
            [circle draw];
        }
    }
    return 0;
}
